torch/csrc/jit/ivalue.h

```diff
@@ -0,0 +1,270 @@
+  }
+  PointerType& operator*() const {
+    return  *get();
+  }
+  PointerType* operator->() const {
+    return get();
+  }
+  operator bool() const {
+    return pImpl != nullptr;
+  }
+private:
+  PointerType * pImpl;
+};
+
+
+template<typename T>
+struct ConstantList;
+struct IValue;
+using Tuple = ConstantList<IValue>;
+using IntList = ConstantList<int64_t>;
+using DoubleList = ConstantList<double>;
+
+// IValue is the generic tagged union used by the interpreter to hold
+// all value types.
+// It is a 16-byte object with an 8-byte payload and an 8-byte tag.
+// The tag is currently 4 bytes to determine the type, and 1 byte
+// to mark whether that type is a subtype of at::Retainable and needs
+// retain/release calls.
+struct IValue {
+  IValue()
+  : IValue(static_cast<int64_t>(0)) {}
+  IValue(const IValue& rhs)
+      : as_int(rhs.as_int),
+        tag(rhs.tag),
+        retainable(rhs.retainable) {
+    if (retainable)
+      as_retainable->retain();
+  }
+  IValue(IValue && rhs) noexcept
+  : IValue() {
+    swap(rhs);
+  }
+  ~IValue() {
+    if (retainable) {
+      as_retainable->release();
+    }
+  }
+  IValue & operator=(IValue && rhs) & {
+    rhs.swap(*this);
+    return *this;
+  }
+  IValue & operator=(IValue const & rhs) & {
+      IValue(rhs).swap(*this);
+      return *this;
+  }
+  void swap(IValue & rhs) {
+    std::swap(as_int, rhs.as_int);
+    std::swap(retainable, rhs.retainable);
+    std::swap(tag, rhs.tag);
+  }
+  // Accessors for subtypes are arragned together below
+  // While some of these accessors could be generated through templates,
+  // we prefer to write them manually for clarity
+
+  // Tensor
+  IValue(at::Tensor t)
+  : tag(Tag::Tensor), retainable(t.defined())  {
+    // note: the undefined tensor is not refcounted, so while it
+    // is tagged as a tensor, retainable is set to false.
+    as_tensor_impl = t.at::detail::TensorBase::detach();
+  }
+  bool isTensor() const { return Tag::Tensor == tag; }
+  at::Tensor toTensor() && {
+    JIT_ASSERT(isTensor());
+    at::Tensor t(as_tensor_impl, /*retain=*/false);
+    clearToZero();
+    return t;
+  }
+  at::Tensor toTensor() const & {
+    JIT_ASSERT(isTensor());
+    return at::Tensor(as_tensor_impl, /*retain=*/true);
+  }
+
+  // Tuple
+  IValue(Shared<Tuple> v);
+  bool isTuple() const { return Tag::Tuple == tag; }
+  Shared<Tuple> toTuple() && {
+    JIT_ASSERT(isTuple());
+    return moveToRetainable<Tuple>();
+  }
+  Shared<Tuple> toTuple() const & {
+    JIT_ASSERT(isTuple());
+    return toRetainable<Tuple>();
+  }
+
+  // Double
+  IValue(double d)
+  : tag(Tag::Double), retainable(false) {
+    as_double = d;
+  }
+  bool isDouble() const { return Tag::Double == tag; }
+  double toDouble() const {
+    JIT_ASSERT(isDouble());
+    return as_double;
+  }
+
+  // Int
+  IValue(int64_t i)
+  : tag(Tag::Int), retainable(false) {
+    as_int = i;
+  }
+  // allow you to pass literals (3, 4) without ambiguity
+  IValue(int32_t i)
```

Comment:
FWIW those literals are `int` literals, and it so only happens in your compiler that it's the same type as `int32_t`. IIRC it's not the case in some weird cases, even on x86_64 systems with 32-bit ints.